### PR TITLE
ENH: review `stats.relfreq` and `stats.cumfreq`; fixes to `stats.histogram`

### LINKED ISSUE #604: Statistics Review: cumfreq (Trac #77)
_Original ticket http://projects.scipy.org/scipy/ticket/77 on 2006-04-02 by @rkern, assigned to @rkern._

The function `cumfreq` in file source:trunk/Lib/stats/stats.py needs review.

Please look over the StatisticsReview guidelines and add your comments below.

### LINKED ISSUE #605: Statistics Review: relfreq (Trac #78)
_Original ticket http://projects.scipy.org/scipy/ticket/78 on 2006-04-02 by @rkern, assigned to @rkern._

The function `relfreq` in file source:trunk/Lib/stats/stats.py needs review.

Please look over the StatisticsReview guidelines and add your comments below.

### PR DESCRIPTION
**Pull request by aeklant**
closes gh-604; closes gh-605

### CODE REVIEW

**Comment by aeklant** [comment]
I did not include the nan check. I will hold off on that until there is a community consensus.

I'm not sure how I would add a plot to the docstring. @rgommers do you have an example on how to do it?

**Comment by rgommers** [comment]
For plotting examples in docstrings, see for example:
http://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.probplot.html#scipy.stats.probplot
http://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.boxcox_normplot.html#scipy.stats.boxcox_normplot

**Comment by aeklant** [comment]
I added a plot to `cumfreq` and `relfreq` in 'examples'. Please review as I'm not sure if the examples are good. 

I plotted the return value; I could have used `pl.hist` but that misses the point of plotting the results of cumulative or relative frequencies, I think.

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -1793,11 +1800,9 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
 
     Examples
     --------
+    >>> import pylab as pl
```
Comment (rgommers) [inline scipy/stats/stats.py]: This should be ``import matplotlib.pyplot as plt``. That's what all other examples do. Pylab is a relic and should be avoided for any new code.

Comment (rgommers) [inline scipy/stats/stats.py]: This doesn't affect the code below though (except for changing ``pl`` to ``plt``).

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -1805,6 +1810,18 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
     >>> extrapoints
     3
 
+    Generate histogram of cumulative frequencies
```
Comment (rgommers) [inline scipy/stats/stats.py]: This needs a blank line between desciption and doctest-format code, or it won't render correctly.

Comment (rgommers) [inline scipy/stats/stats.py]: and same a few lines lower

File: scipy/stats/stats.py [hunk 3]
```diff
@@ -1805,6 +1810,18 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+    >>> cumfreqs, lowlim, binsize, extrapoints = stats.cumfreq(x)
```
Comment (rgommers) [inline scipy/stats/stats.py]: Why not use the namedtuple return here?

File: scipy/stats/stats.py [hunk 4]
```diff
@@ -1805,6 +1810,18 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+    >>> cumfreqs
+    array([ 3.,  3.,  3.,  4.,  4.,  4.,  5.,  5.,  5.,  6. ])
+
+    Plot cumulative frequencies
+    >>> a = np.arange(1, 11)
```
Comment (rgommers) [inline scipy/stats/stats.py]: This falls a bit out of the sky. Can you construct the horizontal axis for your plot from the ``lowlim`` and ``binsize`` return values?

**Comment by rgommers** [comment]
@aeklant the examples are a bit too artificial I think. Here is an example that better illustrates what a cumulative histogram is:

    import numpy as np
    from scipy import stats
    import matplotlib.pyplot as plt


    rng = np.random.RandomState(seed=12345)
    samples = stats.norm.rvs(size=1000, random_state=rng)
    res = stats.cumfreq(samples, numbins=25)

    x = res.lowerlimit + np.linspace(0, res.binsize*res.cumcount.size,
                                     res.cumcount.size)

    fig = plt.figure(figsize=(10, 4))
    ax1 = fig.add_subplot(1, 2, 1)
    ax2 = fig.add_subplot(1, 2, 2)
    ax1.hist(samples, bins=25)
    ax2.bar(x, res.cumcount, width=res.binsize)
    ax1.set_title('Histogram')
    ax2.set_title('Cumulative histogram')
    ax2.set_xlim([x.min(), x.max()])
    plt.show()

Can you see if that all makes sense to you, and then create a similar one for ``relfreq``?

**Comment by aeklant** [comment]
@rgommers ping. Are the new examples good? I kept your example for `cumfreq` and adapted it a little for `relfreq`.

File: scipy/stats/stats.py [hunk 5]
```diff
@@ -1793,18 +1800,41 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+    Calculate cumulative frequencies
+
+    >>> res = stats.cumfreq(samples, numbins=25)
+
+    Calculate space of values for x
+
+    >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.cumcount.size,
+                                         res.cumcount.size)
+
+    Plot 'Histogram' and 'Cumulative histogram'
```
Comment (rgommers) [inline scipy/stats/stats.py]: This needs a blank line below it to render correctly. Also I would format it as ``Plot histogram and cumulative histogram:``

File: scipy/stats/stats.py [hunk 6]
```diff
@@ -1846,17 +1879,43 @@ def relfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
     --------
+    >>> import matplotlib.pyplot as plt
     >>> from scipy import stats
-    >>> a = np.array([1, 4, 2, 1, 3, 1])
-    >>> relfreqs, lowlim, binsize, extrapoints = stats.relfreq(a, numbins=4)
-    >>> relfreqs
+    >>> a = np.array([2, 4, 1, 2, 3, 2])
+    >>> res = stats.relfreq(a, numbins=4)
+    >>> res.frequency
     array([ 0.5       ,  0.16666667,  0.16666667,  0.16666667])
```
Comment (rgommers) [inline scipy/stats/stats.py]: This is incorrect output, should be ``array([ 0.16666667,  0.5       ,  0.16666667,  0.16666667])``

File: scipy/stats/stats.py [hunk 7]
```diff
@@ -1846,17 +1879,43 @@ def relfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
-    >>> np.sum(relfreqs)  # relative frequencies should add up to 1
+    >>> np.sum(res.frequency)  # relative frequencies should add up to 1
     0.99999999999999989
```
Comment (rgommers) [inline scipy/stats/stats.py]: Should be ``1.0``

File: scipy/stats/stats.py [hunk 8]
```diff
@@ -1846,17 +1879,43 @@ def relfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+    Calculate cumulative frequencies
+
+    >>> res = stats.relfreq(samples, numbins=25)
+
+    Calculate space of values for x
+
+    >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.frequency.size,
+                                         res.frequency.size)
+
+    Plot 'Histogram' and 'Cumulative histogram'
```
Comment (rgommers) [inline scipy/stats/stats.py]: This text doesn't match the plot below. You're only plotting one thing, a relative frequency histogram

File: scipy/stats/stats.py [hunk 9]
```diff
@@ -1846,17 +1879,43 @@ def relfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+
+    >>> fig = plt.figure(figsize=(5, 4))
+    >>> ax = fig.add_subplot(1, 1, 1)
+    >>> ax.bar(x, res.frequency, width=res.binsize)
+    >>> ax.set_title('Frequency histogram')
```
Comment (rgommers) [inline scipy/stats/stats.py]: I'd also call it relative frequency histogram here

**Comment by rgommers** [comment]
Make a few comments on details, other than that this is ready to go in.

File: scipy/stats/stats.py [hunk 10]
```diff
@@ -1793,18 +1800,41 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+    >>> samples = stats.norm.rvs(size=1000, random_state=rng)
+
+    Calculate cumulative frequencies
+
+    >>> res = stats.cumfreq(samples, numbins=25)
+
+    Calculate space of values for x
+
+    >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.cumcount.size,
+                                         res.cumcount.size)
```
Comment (ev-br) [inline scipy/stats/stats.py]: Continuation lines should better start with `...`, otherwise itso

**Comment by aeklant** [comment]
I made the changes I made a mistake with the commit. Long story short, I think I should rebase before this PR is merged, for clarity.

File: scipy/stats/stats.py [hunk 11]
```diff
@@ -1846,17 +1880,43 @@ def relfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+    >>> samples = stats.norm.rvs(size=1000, random_state=rng)
+
+    Calculate relative frequencies
+
+    >>> res = stats.relfreq(samples, numbins=25)
+
+    Calculate space of values for x
+
+    >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.frequency.size,
+    ... res.frequency.size)
```
Comment (rgommers) [inline scipy/stats/stats.py]: the ``res.`` has to line up with the 0

File: scipy/stats/stats.py [hunk 12]
```diff
@@ -1793,18 +1800,42 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
...
+    >>> samples = stats.norm.rvs(size=1000, random_state=rng)
+
+    Calculate cumulative frequencies
+
+    >>> res = stats.cumfreq(samples, numbins=25)
+
+    Calculate space of values for x
+
+    >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.cumcount.size,
+    ... res.cumcount.size)
```
Comment (rgommers) [inline scipy/stats/stats.py]: same here, line up lines

**Comment by aeklant** [comment]
Let me know if there are more comments before I rebase.

**Comment by rgommers** [comment]
@aeklant no more comments, please rebase. Would be good to squash this PR down to 2 commits - one for empty input and one for docstring examples.

**Comment by aeklant** [comment]
I have done the rebase and iI think it looks good. Please give one last check to make sure it's merge-ready.

**Comment by rgommers** [comment]
LGTM, merged. Thanks @aeklant

**Pull request merged**